### DISCUSSION TITLE
Prevent race conditions when testing with playwright

### DISCUSSION THREAD

**Discussion by languitar** [bigskysoftware/htmx#2360] [Q&A] [answered]
We have a project using htmx that is tested with playwright. In that scenario, for instance, with boosted links, we see that in playwright htmx actions sometimes don't happen. Things work if the headless browser interaction is slowed down. As far as I could debug this, we end up in a similar situation to what is already [described in the playwright hydration documentation](https://playwright.dev/docs/navigations#hydration): interactions seem to happen before htmx has installed all required callbacks. Is there a common pattern how to solve this with the htmx + playwright combination? Especially for boosting, where the DOM element to interact on is already a link, it feels weird that in some loading state the boosting does not work correctly sometimes.

**Comment by Telroshan** [comment, marked as answer]
I'm also using playwright for htmx-based tests, though I'm not using boosting so I may be off the mark in this situation, let me know.
I have 2 "tricks"
- Check that htmx has been initialized before moving on with the actual tests
```js
const htmx = await page.evaluate("window.htmx")
expect(htmx).not.toEqual(null)
```
You could use this in a `beforeEach` hook for example to avoid repetition
- Define a simple utility function to ensure things were settled between 2 actions where the second action would require the first one to be completed before moving on, such as
```js
async function htmxReady(page: Page) {
    await expect(page.locator('.htmx-request, .htmx-settling, .htmx-swapping, .htmx-added')).toHaveCount(0)
}
```
Which makes sure no request is still in flight (simple assumption that if no element has any of the htmx requests, it's that the requests have all been resolved. Note that if you wanted the most reliable and precise setup, you would use event listeners bound to htmx's lifecycle events). Using it as
```js
await htmxReady(page)
```

**Reply by languitar** [reply]
Thanks, we'll give that a shot and see if that resolves all issues we see.

**Reply by bodograumann** [reply]
Cool ideas. How about writing the first one as `page.waitForFunction(() => window.htmx)`? This should wait for htmx to be there.
